### DISCUSSION TITLE
Running Tests in Folders Concurrently, but Folders in Parallel

### DISCUSSION THREAD

**Discussion by kevinswarner** [vitest-dev/vitest#6796] [Q&A]
We have a project where we have thousands of test files. We cannot run them in parallel by default because they operate against a shared database and the tests clobber each other and result in inconsistent results. We have tests organized into different folders.

For example...

```
src/module1
  foo.test.ts
  foo.ts
  bar.test.ts
  bar.ts
src/module2
  {MULTIPLE FILES}
src/module3
  {MULTIPLE FILES}
```

Our goal is to run each test within a folder sequentially, but run tests for the folders in parallel. We are able to make sure the tests in each different folder operate against separate databases.

We have tried to use the "concurrently" package to simply execute "vitest run" for different folders in parallel, but it will not work as expected. The folders still run sequentially.

I attempted using a custom sequencer, but had no luck with that.

If anyone has faced this problem, I appreciate any guidance.

Thanks!

**Comment by hi-ogawa** [comment]
> We have tried to use the "concurrently" package to simply execute "vitest run" for different folders in parallel, but it will not work as expected. The folders still run sequentially.

This sounds like only a matter of scripting and I'm not sure Vitest is relevant. I would imagine something like this would work (though not sure about concurrently's syntax or package manager).

```
{
  "scripts": {
    "test": "concurrently 'pnpm test:module1' 'pnpm test:module2'",
    "test:module1": "vitest src/module1/ --no-file-parallelism",
    "test:module2": "vitest src/module2/ --no-file-parallelism"
  }
}
```

You can also move `--no-file-parallelism` to config as `fileParallelism: false` https://vitest.dev/config/#fileparallelism
